# Link de repositorio #

https://github.com/Saiyan-Javi/Lab8DS

In [66]:
import pandas as pd


In [67]:
df2014 = pd.read_excel('Vehiculos 2014.xlsx')
df2015 = pd.read_excel('Vehiculos 2015.xlsx')  
df2016 = pd.read_excel('Vehiculos 2016.xlsx')
df2017 = pd.read_excel('Vehiculos 2017.xlsx')
df2018 = pd.read_excel('Vehiculos 2018.xlsx')
df2019 = pd.read_excel('Vehiculos 2019.xlsx')
df2020 = pd.read_excel('Vehiculos 2020.xlsx')
df2021 = pd.read_excel('Vehiculos 2021.xlsx')
df2022 = pd.read_excel('Vehiculos 2022.xlsx')
df2023 = pd.read_excel('Vehiculos 2023.xlsx')



c:\spark_lab\venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")
c:\spark_lab\venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [68]:
df2014.drop("num_correlativo", axis=1, inplace=True)
df2014.drop("corre_base", axis=1, inplace=True)
df2014["año_ocu"] = 2014 

In [69]:
df2015.drop("núm_corre", axis=1, inplace=True)
df2015["año_ocu"] = 2015

In [70]:
df2016.drop("núm_corre", axis=1, inplace=True)
df2016["año_ocu"] = 2016

In [71]:
df2017.drop("núm_corre", axis=1, inplace=True)
df2018.drop("núm_corre", axis=1, inplace=True)
df2019.drop("núm_corre", axis=1, inplace=True)
df2020.drop("Núm_corre", axis=1, inplace=True)
df2021.drop("núm_corre", axis=1, inplace=True)
df2022.drop("núm_corre", axis=1, inplace=True)
df2023.drop("núm_corre", axis=1, inplace=True)


In [72]:
import unicodedata

def normalizar_columnas(df):
    df.columns = [
        ''.join(
            c for c in unicodedata.normalize('NFKD', col)
            if not unicodedata.combining(c)
        ).lower()
        for col in df.columns
    ]
    return df

df2014 = normalizar_columnas(df2014)
df2015 = normalizar_columnas(df2015)
df2016 = normalizar_columnas(df2016)
df2017 = normalizar_columnas(df2017)
df2018 = normalizar_columnas(df2018)
df2019 = normalizar_columnas(df2019)
df2020 = normalizar_columnas(df2020)
df2021 = normalizar_columnas(df2021)
df2022 = normalizar_columnas(df2022)
df2023 = normalizar_columnas(df2023)

In [73]:
df2014 = df2014[sorted(df2014.columns)]
df2015 = df2015[sorted(df2015.columns)]
df2016 = df2016[sorted(df2016.columns)]
df2017 = df2017[sorted(df2017.columns)]
df2018 = df2018[sorted(df2018.columns)]
df2019 = df2019[sorted(df2019.columns)]
df2020 = df2020[sorted(df2020.columns)]
df2021 = df2021[sorted(df2021.columns)]
df2022 = df2022[sorted(df2022.columns)]
df2023 = df2023[sorted(df2023.columns)]

In [74]:
print("2014:", df2014.columns.tolist())
print("2015:", df2015.columns.tolist())
print("2016:", df2016.columns.tolist())
print("2017:", df2017.columns.tolist())
print("2018:", df2018.columns.tolist())
print("2019:", df2019.columns.tolist())
print("2020:", df2020.columns.tolist())
print("2021:", df2021.columns.tolist())
print("2022:", df2022.columns.tolist())
print("2023:", df2023.columns.tolist())

2014: ['ano_ocu', 'area_geo_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_con', 'estado_con', 'g_edad', 'g_hora', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_con', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2015: ['ano_ocu', 'area_geo_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per', 'edad_quinquenales', 'estado_con', 'g_edad_60ymas', 'g_edad_80ymas', 'g_hora', 'g_hora_5', 'g_modelo_veh', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_per', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2016: ['ano_ocu', 'area_geo_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per', 'edad_quinquenales', 'estado_con', 'g_edad_60ymas', 'g_edad_80ymas', 'g_hora', 'g_hora_5', 'g_modelo_veh', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_per', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2017: ['ano_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per'

In [75]:
df2014.drop("area_geo_ocu", axis=1, inplace=True)
df2015.drop("area_geo_ocu", axis=1, inplace=True)
df2016.drop("area_geo_ocu", axis=1, inplace=True)

In [76]:
# Cambiar el nombre de la columna 'edad_con' a 'edad_per' en df2014
df2014.rename(columns={'edad_con': 'edad_per'}, inplace=True)

# Crear la columna 'edad_quinquenales' en base a 'edad_per'
def calcular_quinquenal(edad):
    try:
        edad = int(edad)
        if 0 <= edad <= 89:
            return (edad // 5) + 1
        elif edad >= 90:
            return 18
        else:
            return 18
    except:
        return 18

df2014['edad_quinquenales'] = df2014['edad_per'].apply(calcular_quinquenal)

In [77]:
# Cambiar el nombre de la columna 'g_edad' a 'g_edad_60ymas' en df2014
df2014.rename(columns={'g_edad': 'g_edad_60ymas'}, inplace=True)

# Crear la columna 'g_edad_80ymas' en base a 'edad_per'
def calcular_g_edad_80ymas(edad):
    try:
        edad = int(edad)
        if edad == 999:
            return 16
        elif edad < 15:
            return 1
        elif 15 <= edad <= 19:
            return 2
        elif 20 <= edad <= 24:
            return 3
        elif 25 <= edad <= 29:
            return 4
        elif 30 <= edad <= 34:
            return 5
        elif 35 <= edad <= 39:
            return 6
        elif 40 <= edad <= 44:
            return 7
        elif 45 <= edad <= 49:
            return 8
        elif 50 <= edad <= 54:
            return 9
        elif 55 <= edad <= 59:
            return 10
        elif 60 <= edad <= 64:
            return 11
        elif 65 <= edad <= 69:
            return 12
        elif 70 <= edad <= 74:
            return 13
        elif 75 <= edad <= 80:
            return 14
        else:
            return 15
    except:
        return None

df2014['g_edad_80ymas'] = df2014['edad_per'].apply(calcular_g_edad_80ymas)

In [78]:
def calcular_g_hora_5(valor):
    if valor in [1, 2]:
        return 1
    elif valor == 3:
        return 2
    elif valor == 4:
        return 3
    else:
        return 4

df2014['g_hora_5'] = df2014['g_hora'].apply(calcular_g_hora_5)

In [79]:
def calcular_g_modelo_veh(modelo):
    try:
        modelo = int(modelo)
        if 1970 <= modelo <= 1979:
            return 1
        elif 1980 <= modelo <= 1989:
            return 2
        elif 1990 <= modelo <= 1999:
            return 3
        elif 2000 <= modelo <= 2009:
            return 4
        elif modelo >= 2010:
            return 5
        else:
            return None
    except:
        return None

df2014['g_modelo_veh'] = df2014['modelo_veh'].apply(calcular_g_modelo_veh)

In [80]:
df2014.rename(columns={'sexo_con': 'sexo_per'}, inplace=True)

In [81]:
df2021.drop('zona_ciudad', axis=1, inplace=True)

In [82]:
df2014 = df2014[sorted(df2014.columns)]
df2015 = df2015[sorted(df2015.columns)]
df2016 = df2016[sorted(df2016.columns)]
df2017 = df2017[sorted(df2017.columns)]
df2018 = df2018[sorted(df2018.columns)]
df2019 = df2019[sorted(df2019.columns)]
df2020 = df2020[sorted(df2020.columns)]
df2021 = df2021[sorted(df2021.columns)]
df2022 = df2022[sorted(df2022.columns)]
df2023 = df2023[sorted(df2023.columns)]

print("2014:", df2014.columns.tolist())
print("2015:", df2015.columns.tolist())
print("2016:", df2016.columns.tolist())
print("2017:", df2017.columns.tolist())
print("2018:", df2018.columns.tolist())
print("2019:", df2019.columns.tolist())
print("2020:", df2020.columns.tolist())
print("2021:", df2021.columns.tolist())
print("2022:", df2022.columns.tolist())
print("2023:", df2023.columns.tolist())

2014: ['ano_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per', 'edad_quinquenales', 'estado_con', 'g_edad_60ymas', 'g_edad_80ymas', 'g_hora', 'g_hora_5', 'g_modelo_veh', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_per', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2015: ['ano_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per', 'edad_quinquenales', 'estado_con', 'g_edad_60ymas', 'g_edad_80ymas', 'g_hora', 'g_hora_5', 'g_modelo_veh', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_per', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2016: ['ano_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 'dia_sem_ocu', 'edad_per', 'edad_quinquenales', 'estado_con', 'g_edad_60ymas', 'g_edad_80ymas', 'g_hora', 'g_hora_5', 'g_modelo_veh', 'hora_ocu', 'marca_veh', 'mayor_menor', 'mes_ocu', 'modelo_veh', 'mupio_ocu', 'sexo_per', 'tipo_eve', 'tipo_veh', 'zona_ocu']
2017: ['ano_ocu', 'color_veh', 'depto_ocu', 'dia_ocu', 

# Analisis por columnas #

In [83]:
# Mostrar valores únicos por columna, agrupando por columna y mostrando los valores por año
columnas = df2014.columns  # Asume que todas las columnas están en todos los dataframes

for col in columnas:
    print(f"\nColumna: {col}")
    for year, df in zip(range(2014, 2024), [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021, df2022, df2023]):
        if col in df.columns:
            print(f"  {year}: {df[col].unique()}")
        else:
            print(f"  {year}: (no existe)")


Columna: ano_ocu
  2014: [2014]
  2015: [2015]
  2016: [2016]
  2017: [2017]
  2018: [2018]
  2019: [2019]
  2020: [2020]
  2021: [2021]
  2022: [2022]
  2023: [2023]

Columna: color_veh
  2014: [ 5 99 11  1  7  2 12  3  4  6  8 10  9 14 16 15]
  2015: [11  5  3  1  6  4  7  9  2 99  8 14 10 16 12 15]
  2016: [11  5  3  1  6  4  7  9  2 99  8 14 10 16 12 15]
  2017: [ 7  2  1  4  5 99  9 10  6  3  8 11 17 14 15 12 16]
  2018: [ 5  2  1  6  9 14  3 11  4  8 99  7 10 15 17 12 16]
  2019: [ 5 14  8  3  1 99  9  2  6  4  7 11 10 16 15 17 12]
  2020: [ 3  4 99  5  6  1  2 15  9 14 11  8  7 10 17 12 16]
  2021: [ 4  1  5 99  2  7  8  3  6  9 11 10 15 17 14 12 16]
  2022: ['Azul' 'Celeste' 'Blanco' 'Ignorado' 'Amarillo' 'Negro' 'Verde' 'Rojo'
 'Gris' 'Turquesa' 'Corinto' 'Café' 'Anaranjado' 'Beige' 'Morado' 'Rosado'
 'Varios colores' 'Marfil']
  2023: ['Negro' 'Gris' 'Azul' 'Blanco' 'Rojo' 'Ignorado' 'Corinto' 'Verde'
 'Amarillo' 'Café' 'Anaranjado' 'Turquesa' 'Beige' 'Morado' 'Celeste'
 'Ro

In [84]:
# Diccionario de mapeo de códigos a colores
colores = {
    1: "Rojo",
    2: "Blanco",
    3: "Azul",
    4: "Gris",
    5: "Negro",
    6: "Verde",
    7: "Amarillo",
    8: "Celeste",
    9: "Corinto",
    10: "Café",
    11: "Beige",
    12: "Turquesa",
    13: "Marfil",
    14: "Anaranjado",
    15: "Aqua",
    16: "Morado",
    17: "Rosado",
    99: "Ignorado"
}

def mapear_color(valor):
    try:
        return colores.get(int(valor), "Otro")
    except:
        return "Otro"

# Aplica la función a todos los dataframes si existe la columna color_veh
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'color_veh' in df.columns:
        df['color_veh'] = df['color_veh'].apply(mapear_color)

In [85]:
# Diccionario de mapeo de códigos a departamentos
departamentos = {
    1: "Guatemala",
    2: "El Progreso",
    3: "Sacatepéquez",
    4: "Chimaltenango",
    5: "Escuintla",
    6: "Santa Rosa",
    7: "Sololá",
    8: "Totonicapán",
    9: "Quetzaltenango",
    10: "Suchitepéquez",
    11: "Retalhuleu",
    12: "San Marcos",
    13: "Huehuetenango",
    14: "Quiché",
    15: "Baja Verapaz",
    16: "Alta Verapaz",
    17: "Petén",
    18: "Izabal",
    19: "Zacapa",
    20: "Chiquimula",
    21: "Jalapa",
    22: "Jutiapa"
}

def mapear_depto(valor):
    try:
        return departamentos.get(int(valor), "Otro")
    except:
        return "Otro"

# Aplica la función a todos los dataframes si existe la columna depto_ocu
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'depto_ocu' in df.columns:
        df['depto_ocu'] = df['depto_ocu'].apply(mapear_depto)

In [86]:
# Cambiar 999 por 'Ignorada' en edad_per para todos los dataframes donde exista la columna
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021, df2022, df2023]:
    if 'edad_per' in df.columns:
        df['edad_per'] = df['edad_per'].replace(999, 'Ignorada')

# Diccionario para dia_sem_ocu
dias_semana = {
    1: "Lunes",
    2: "Martes",
    3: "Miercoles",
    4: "Jueves",
    5: "Viernes",
    6: "Sabado",
    7: "Domingo"
}

# Aplica el mapeo para todos los dataframes
for year, df in zip(range(2014, 2024), [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]):
    if 'dia_sem_ocu' in df.columns:
        df['dia_sem_ocu'] = df['dia_sem_ocu'].map(dias_semana)

In [87]:
if 'dia_sem_ocu' in df2023.columns:
        df['dia_sem_ocu'] = df['dia_sem_ocu'].replace('Miércoles', 'Miercoles')
        df['dia_sem_ocu'] = df['dia_sem_ocu'].replace('Sábado', 'Sabado')

In [88]:
# Crear un diccionario para mapear el número de quinquenal al intervalo de años
quinquenal_intervalos = {
    1: '0 - 4',
    2: '5 - 9',
    3: '10 - 14',
    4: '15 - 19',
    5: '20 - 24',
    6: '25 - 29',
    7: '30 - 34',
    8: '35 - 39',
    9: '40 - 44',
    10: '45 - 49',
    11: '50 - 54',
    12: '55 - 59',
    13: '60 - 64',
    14: '65 - 69',
    15: '70 - 74',
    16: '75 - 79',
    17: '80 o mas',
    18: 'Ignorado'
}

# Aplica el mapeo en la columna 'edad_quinquenales' de df2014 a df2021
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'edad_quinquenales' in df.columns:
        df['edad_quinquenales'] = df['edad_quinquenales'].map(quinquenal_intervalos).fillna(df['edad_quinquenales'])

In [89]:
if 'edad_quinquenales' in df2023.columns:
    df2023['edad_quinquenales'] = df2023['edad_quinquenales'].replace('80 o m?s', '80 o mas')

if 'edad_quinquenales' in df2022.columns:
    df2022['edad_quinquenales'] = df2022['edad_quinquenales'].replace('80 o más', '80 o mas')

In [90]:
# Diccionario de mapeo para estado_con
estado_con_dict = {
    1: 'No Ebrio',
    2: 'Ebrio',
    9: 'Ignorado'
}

for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'estado_con' in df.columns:
        df['estado_con'] = df['estado_con'].replace(estado_con_dict)

In [91]:
mupio_dict = {
    101: "Guatemala",
    102: "Santa Catarina Pinula",
    103: "San José Pinula",
    104: "San José del Golfo",
    105: "Palencia",
    106: "Chinautla",
    107: "San Pedro Ayampuc",
    108: "Mixco",
    109: "San Pedro Sacatepéquez",
    110: "San Juan Sacatepéquez",
    111: "San Raymundo",
    112: "Chuarrancho",
    113: "Fraijanes",
    114: "Amatitlán",
    115: "Villa Nueva",
    116: "Villa Canales",
    117: "Petapa",
    201: "Guastatoya",
    202: "Morazán",
    203: "San Agustín Acasaguastlán",
    204: "San Cristóbal Acasaguastlán",
    205: "El Jícaro",
    206: "Sansare",
    207: "Sanarate",
    208: "San Antonio la Paz",
    301: "Antigua Guatemala",
    302: "Jocotenango",
    303: "Pastores",
    304: "Sumpango",
    305: "Santo Domingo Xenacoj",
    306: "Santiago Sacatepéquez",
    307: "San Bartolomé Milpas Altas",
    308: "San Lucas Sacatepéquez",
    309: "Santa Lucía Milpas Altas",
    310: "Magdalena Milpas Altas",
    311: "Santa María de Jesús",
    312: "Ciudad Vieja",
    313: "San Miguel Dueñas",
    314: "San Juan Alotenango",
    315: "San Antonio Aguas Calientes",
    316: "Santa Catarina Barahona",
    401: "Chimaltenango",
    402: "San José Poaquíl",
    403: "San Martín Jilotepeque",
    404: "Comalapa",
    405: "Santa Apolonia",
    406: "Tecpán Guatemala",
    407: "Patzún",
    408: "San Miguel Pochuta",
    409: "Patzicía",
    410: "Santa Cruz Balanyá",
    411: "Acatenango",
    412: "Yepocapa",
    413: "San Andrés Itzapa",
    414: "Parramos",
    415: "Zaragoza",
    416: "El Tejar",
    501: "Escuintla",
    502: "Santa Lucía Cotzumalguapa",
    503: "La Democracia",
    504: "Siquinalá",
    505: "Masagua",
    506: "Tiquisate",
    507: "La Gomera",
    508: "Guanagazapa",
    509: "San José",
    510: "Iztapa",
    511: "Palín",
    512: "San Vicente Pacaya",
    513: "Nueva Concepción",
    514: "Sipacate",
    601: "Cuilapa",
    602: "Barberena",
    603: "Santa Rosa de Lima",
    604: "Casillas",
    605: "San Rafael Las Flores",
    606: "Oratorio",
    607: "San Juan Tecuaco",
    608: "Chiquimulilla",
    609: "Taxisco",
    610: "Santa María Ixhuatán",
    611: "Guazacapán",
    612: "Santa Cruz el Naranjo",
    613: "Pueblo Nuevo Viñas",
    614: "Nueva Santa Rosa",
    701: "Sololá",
    702: "San José Chacayá",
    703: "Santa María Visitación",
    704: "Santa Lucía Utatlán",
    705: "Nahualá",
    706: "Santa Catarina Ixtahuacán",
    707: "Santa Clara La Laguna",
    708: "Concepción",
    709: "San Andrés Semetabaj",
    710: "Panajachel",
    711: "Santa Catarina Palopó",
    712: "San Antonio Palopó",
    713: "San Lucas Tolimán",
    714: "Santa Cruz La Laguna",
    715: "San Pablo La Laguna",
    716: "San Marcos La Laguna",
    717: "San Juan La Laguna",
    718: "San Pedro La Laguna",
    719: "Santiago Atitlán",
    801: "Totonicapán",
    802: "San Cristóbal Totonicapán",
    803: "San Francisco El Alto",
    804: "San Andrés Xecul",
    805: "Momostenango",
    806: "Santa María Chiquimula",
    807: "Santa Lucía La Reforma",
    808: "San Bartolo Aguas Calientes",
    901: "Quetzaltenango",
    902: "Salcajá",
    903: "Olintepeque",
    904: "San Carlos Sija",
    905: "Sibilia",
    906: "Cabricán",
    907: "Cajolá",
    908: "San Miguel Siguilá",
    909: "San Juan Ostuncalco",
    910: "San Mateo",
    911: "Concepción Chiquirichapa",
    912: "San Martín Sacatepéquez",
    913: "Almolonga",
    914: "Cantel",
    915: "Huitán",
    916: "Zunil",
    917: "Colomba",
    918: "San Francisco La Unión",
    919: "El Palmar",
    920: "Coatepeque",
    921: "Génova",
    922: "Flores Costa Cuca",
    923: "La Esperanza",
    924: "Palestina de Los Altos",
    1001: "Mazatenango",
    1002: "Cuyotenango",
    1003: "San Francisco Zapotitlán",
    1004: "San Bernardino",
    1005: "San José El Idolo",
    1006: "Santo Domingo Suchitepéquez",
    1007: "San Lorenzo",
    1008: "Samayac",
    1009: "San Pablo Jocopilas",
    1010: "San Antonio Suchitepéquez",
    1011: "San Miguel Panán",
    1012: "San Gabriel",
    1013: "Chicacao",
    1014: "Patulul",
    1015: "Santa Bárbara",
    1016: "San Juan Bautista",
    1017: "Santo Tomás La Unión",
    1018: "Zunilito",
    1019: "Pueblo Nuevo",
    1020: "Río Bravo",
    1021: "San José La Máquina",
    1101: "Retalhuleu",
    1102: "San Sebastián",
    1103: "Santa Cruz Muluá",
    1104: "San Martín Zapotitlán",
    1105: "San Felipe Retalhuleu",
    1106: "San Andrés Villa Seca",
    1107: "Champerico",
    1108: "Nuevo San Carlos",
    1109: "El Asintal",
    1201: "San Marcos",
    1202: "San Pedro Sacatepéquez",
    1203: "San Antonio Sacatepéquez",
    1204: "Comitancillo",
    1205: "San Miguel Ixtahuacán",
    1206: "Concepción Tutuapa",
    1207: "Tacaná",
    1208: "Sibinal",
    1209: "Tajumulco",
    1210: "Tejutla",
    1211: "San Rafael Pie de la Cuesta",
    1212: "Nuevo Progreso",
    1213: "El Tumbador",
    1214: "El Rodeo",
    1215: "Malacatán",
    1216: "Catarina",
    1217: "Ayutla",
    1218: "Ocós",
    1219: "San Pablo",
    1220: "El Quetzal",
    1221: "La Reforma",
    1222: "Pajapita",
    1223: "Ixchiguán",
    1224: "San José Otenam",
    1225: "San Cristóbal Cucho",
    1226: "Sipacapa",
    1227: "Esquipulas Palo Gordo",
    1228: "Río Blanco",
    1229: "San Lorenzo",
    1230: "La Blanca",
    1301: "Huehuetenango",
    1302: "Chiantla",
    1303: "Malacatancito",
    1304: "Cuilco",
    1305: "Nentón",
    1306: "San Pedro Necta",
    1307: "Jacaltenango",
    1308: "San Pedro Soloma",
    1309: "San Ildefonso Ixtahuacán",
    1310: "Santa Bárbara",
    1311: "La Libertad",
    1312: "La Democracia",
    1313: "San Miguel Acatán",
    1314: "San Rafael La Independencia",
    1315: "Todos Santos Cuchumatán",
    1316: "San Juan Atitán",
    1317: "Santa Eulalia",
    1318: "San Mateo Ixtatán",
    1319: "Colotenango",
    1320: "San Sebastián Huehuetenango",
    1321: "Tectitán",
    1322: "Concepción Huista",
    1323: "San Juan Ixcoy",
    1324: "San Antonio Huista",
    1325: "San Sebastián Coatán",
    1326: "Santa Cruz Barillas",
    1327: "Aguacatán",
    1328: "San Rafael Petzal",
    1329: "San Gaspar Ixchil",
    1330: "Santiago Chimaltenango",
    1331: "Santa Ana Huista",
    1332: "Unión Cantinil",
    1333: "Petatán",
    1401: "Santa Cruz del Quiché",
    1402: "Chiché",
    1403: "Chinique",
    1404: "Zacualpa",
    1405: "Chajul",
    1406: "Chichicastenango",
    1407: "Patzité",
    1408: "San Antonio Ilotenango",
    1409: "San Pedro Jocopilas",
    1410: "Cunén",
    1411: "San Juan Cotzal",
    1412: "Joyabaj",
    1413: "Santa María Nebaj",
    1414: "San Andrés Sajcabajá",
    1415: "San Miguel Uspantán",
    1416: "Sacapulas",
    1417: "San Bartolomé Jocotenango",
    1418: "Canillá",
    1419: "Chicamán",
    1420: "Ixcán",
    1421: "Pachulum",
    1501: "Salamá",
    1502: "San Miguel Chicaj",
    1503: "Rabinal",
    1504: "Cubulco",
    1505: "Granados",
    1506: "El Chol",
    1507: "San Jerónimo",
    1508: "Purulhá",
    1601: "Cobán",
    1602: "Santa Cruz Verapaz",
    1603: "San Cristóbal Verapaz",
    1604: "Tactic",
    1605: "Tamahú",
    1606: "Tucurú",
    1607: "Panzós",
    1608: "Senahú",
    1609: "San Pedro Carchá",
    1610: "San Juan Chamelco",
    1611: "Lanquín",
    1612: "Cahabón",
    1613: "Chisec",
    1614: "Chahal",
    1615: "Fray Bartolomé de las Casas",
    1616: "Santa Caterina La Tinta",
    1617: "Raxruhá",
    1701: "Flores",
    1702: "San José",
    1703: "San Benito",
    1704: "San Andrés",
    1705: "La Libertad",
    1706: "San Francisco",
    1707: "Santa Ana",
    1708: "Dolores",
    1709: "San Luis",
    1710: "Sayaxché",
    1711: "Melchor de Mencos",
    1712: "Poptún",
    1713: "Las Cruces",
    1714: "El Chal",
    1801: "Puerto Barrios",
    1802: "Livingston",
    1803: "El Estor",
    1804: "Morales",
    1805: "Los Amates",
    1901: "Zacapa",
    1902: "Estanzuela",
    1903: "Río Hondo",
    1904: "Gualán",
    1905: "Teculután",
    1906: "Usumatlán",
    1907: "Cabañas",
    1908: "San Diego",
    1909: "La Unión",
    1910: "Huité",
    1911: "San Jorge",
    2001: "Chiquimula",
    2002: "San José La Arada",
    2003: "San Juan Ermita",
    2004: "Jocotán",
    2005: "Camotán",
    2006: "Olopa",
    2007: "Esquipulas",
    2008: "Concepción Las Minas",
    2009: "Quezaltepeque",
    2010: "San Jacinto",
    2011: "Ipala",
    2101: "Jalapa",
    2102: "San Pedro Pinula",
    2103: "San Luis Jilotepeque",
    2104: "San Manuel Chaparrón",
    2105: "San  Carlos Alzatate",
    2106: "Monjas",
    2107: "Mataquescuintla",
    2201: "Jutiapa",
    2202: "El Progreso",
    2203: "Santa Catarina Mita",
    2204: "Agua Blanca",
    2205: "Asunción Mita",
    2206: "Yupiltepeque",
    2207: "Atescatempa",
    2208: "Jerez",
    2209: "El Adelanto",
    2210: "Zapotitlán",
    2211: "Comapa",
    2212: "Jalpatagua",
    2213: "Conguaco",
    2214: "Moyuta",
    2215: "Pasaco",
    2216: "San José Acatempa",
    2217: "Quesada",
}

# Aplica el mapeo de claves a nombres en la columna 'mupio_ocu' del 2014 al 2021
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'mupio_ocu' in df.columns:
        df['mupio_ocu'] = df['mupio_ocu'].replace(mupio_dict)


In [92]:
for col in columnas:
    print(f"\nColumna: {col}")
    for year, df in zip(range(2014, 2024), [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021, df2022, df2023]):
        if col in df.columns:
            print(f"  {year}: {df[col].unique()}")
        else:
            print(f"  {year}: (no existe)")


Columna: ano_ocu
  2014: [2014]
  2015: [2015]
  2016: [2016]
  2017: [2017]
  2018: [2018]
  2019: [2019]
  2020: [2020]
  2021: [2021]
  2022: [2022]
  2023: [2023]

Columna: color_veh
  2014: ['Negro' 'Ignorado' 'Beige' 'Rojo' 'Amarillo' 'Blanco' 'Turquesa' 'Azul'
 'Gris' 'Verde' 'Celeste' 'Café' 'Corinto' 'Anaranjado' 'Morado' 'Aqua']
  2015: ['Beige' 'Negro' 'Azul' 'Rojo' 'Verde' 'Gris' 'Amarillo' 'Corinto'
 'Blanco' 'Ignorado' 'Celeste' 'Anaranjado' 'Café' 'Morado' 'Turquesa'
 'Aqua']
  2016: ['Beige' 'Negro' 'Azul' 'Rojo' 'Verde' 'Gris' 'Amarillo' 'Corinto'
 'Blanco' 'Ignorado' 'Celeste' 'Anaranjado' 'Café' 'Morado' 'Turquesa'
 'Aqua']
  2017: ['Amarillo' 'Blanco' 'Rojo' 'Gris' 'Negro' 'Ignorado' 'Corinto' 'Café'
 'Verde' 'Azul' 'Celeste' 'Beige' 'Rosado' 'Anaranjado' 'Aqua' 'Turquesa'
 'Morado']
  2018: ['Negro' 'Blanco' 'Rojo' 'Verde' 'Corinto' 'Anaranjado' 'Azul' 'Beige'
 'Gris' 'Celeste' 'Ignorado' 'Amarillo' 'Café' 'Aqua' 'Rosado' 'Turquesa'
 'Morado']
  2019: ['Negro' 'An

In [93]:
if 'hora_ocu' in df2021.columns:
    df2021['hora_ocu'] = df2021['hora_ocu'].replace(99, 'Ignorada')

In [94]:
# Diccionario de mapeo para los meses
meses = {
    1: "Enero",
    2: "Febrero",
    3: "Marzo",
    4: "Abril",
    5: "Mayo",
    6: "Junio",
    7: "Julio",
    8: "Agosto",
    9: "Septiembre",
    10: "Octubre",
    11: "Noviembre",
    12: "Diciembre"
}

# Aplica el mapeo para todos los dataframes donde exista la columna mes_ocu
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'mes_ocu' in df.columns:
        df['mes_ocu'] = df['mes_ocu'].replace(meses)

In [95]:
# Diccionario de mapeo para sexo_per
sexo_dict = {
    1: 'Hombre',
    2: 'Mujer',
    9: 'Ignorado'
}

# Aplica el mapeo para todos los dataframes donde exista la columna sexo_per del 2014 al 2021
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'sexo_per' in df.columns:
        df['sexo_per'] = df['sexo_per'].replace(sexo_dict)

In [96]:
# Diccionario de mapeo para los tipos de evento
eventos = {
    1: "Colisión",
    2: "Choque",
    3: "Vuelco",
    4: "Caída",
    5: "Atropello",
    6: "Derrape",
    7: "Embarrancó",
    8: "Encunetó",
    99: "Ignorado"
}

# Aplica el mapeo para todos los dataframes donde exista la columna mes_ocu
for df in [df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021]:
    if 'tipo_eve' in df.columns:
        df['tipo_eve'] = df['tipo_eve'].replace(eventos)

In [97]:
# Concatenar todos los DataFrames
df_total = pd.concat([df2014, df2015, df2016, df2017, df2018, df2019, df2020, df2021, df2022, df2023], ignore_index=True)

# Guardar como archivo Excel en la carpeta actual
df_total.to_excel("Vehiculos_concatenado.xlsx", index=False)

# Analisis #

In [98]:
import os
import sys

# Fuerza a Spark a usar el Python de este mismo entorno
os.environ["PYSPARK_PYTHON"] = sys.executable
os.environ["PYSPARK_DRIVER_PYTHON"] = sys.executable

In [99]:
from pyspark.sql import SparkSession

spark = (
    SparkSession.builder
    .appName("MiApp")
    .getOrCreate()
)
print("Spark runtime:", spark.version)

Spark runtime: 3.5.1


Dataframe de vehiculos

In [100]:
pdfveh = pd.read_excel("Vehiculos_concatenado.xlsx")
pdfveh = pdfveh.astype(str)
dfveh = spark.createDataFrame(pdfveh)

dfveh.show()


+-------+---------+-------------+-------+-----------+--------+-----------------+----------+-------------+-------------+------+--------+------------+--------+---------+-----------+-------+----------+--------------------+--------+--------+--------+--------+
|ano_ocu|color_veh|    depto_ocu|dia_ocu|dia_sem_ocu|edad_per|edad_quinquenales|estado_con|g_edad_60ymas|g_edad_80ymas|g_hora|g_hora_5|g_modelo_veh|hora_ocu|marca_veh|mayor_menor|mes_ocu|modelo_veh|           mupio_ocu|sexo_per|tipo_eve|tipo_veh|zona_ocu|
+-------+---------+-------------+-------+-----------+--------+-----------------+----------+-------------+-------------+------+--------+------------+--------+---------+-----------+-------+----------+--------------------+--------+--------+--------+--------+
|   2014|    Negro|    Guatemala|      1|  Miercoles|      20|          20 - 24|  Ignorado|            3|            3|     1|       1|         nan|       4|       32|          1|  Enero|         3|           Guatemala|  Hombre|Coli

In [108]:
num_filas = dfveh.count()
num_filas

103870

Hay 103870 vehiculos involucrados.

Dataframe fallecidos y lesionados

In [101]:
pdffal = pd.read_csv("fallecimientos_2014_2023_estandarizado2.csv")
pdffal = pdffal.astype(str)
dffal = spark.createDataFrame(pdffal)

dffal.show()


C:\Users\ASUS TUF\AppData\Local\Temp\ipykernel_27984\2247692950.py:1: DtypeWarning: Columns (2,4,11,14,15,19,20,22,23) have mixed types. Specify dtype option on import or set low_memory=False.
  pdffal = pd.read_csv("fallecimientos_2014_2023_estandarizado2.csv")


+-------+-------+--------+-------------+--------+-------+-----------+------------+--------------------+--------+--------+--------+-------------+-------------+-----------------+-----------+-----------+---------+---------+----------+------------+--------+---------+-----------+
|ano_ocu|dia_ocu|hora_ocu|       g_hora|g_hora_5|mes_ocu|dia_sem_ocu|   depto_ocu|           mupio_ocu|zona_ocu|sexo_per|edad_per|g_edad_80ymas|g_edad_60ymas|edad_quinquenales|mayor_menor|   tipo_veh|marca_veh|color_veh|modelo_veh|g_modelo_veh|tipo_eve| fall_les|int_o_noint|
+-------+-------+--------+-------------+--------+-------+-----------+------------+--------------------+--------+--------+--------+-------------+-------------+-----------------+-----------+-----------+---------+---------+----------+------------+--------+---------+-----------+
|    nan|     18|       2|00:00 a 05:59|     nan|  Enero|     Sábado|   Guatemala|           Guatemala|      18|  Hombre|Ignorado|          nan|          nan|         Ignor

In [109]:
num_filas = dffal.count()
num_filas

100357

Hay 100357 fallecidos/lesionados en los accidentes.

Dataframe hechos 

In [103]:
pdfhecho = pd.read_csv("fallecimientos_2014_2023_estandarizado2.csv")
pdfhecho = pdfhecho.astype(str)
dfhecho = spark.createDataFrame(pdfhecho)

dfhecho.show()

C:\Users\ASUS TUF\AppData\Local\Temp\ipykernel_27984\1027119046.py:1: DtypeWarning: Columns (2,4,11,14,15,19,20,22,23) have mixed types. Specify dtype option on import or set low_memory=False.
  pdfhecho = pd.read_csv("fallecimientos_2014_2023_estandarizado2.csv")


+-------+-------+--------+-------------+--------+-------+-----------+------------+--------------------+--------+--------+--------+-------------+-------------+-----------------+-----------+-----------+---------+---------+----------+------------+--------+---------+-----------+
|ano_ocu|dia_ocu|hora_ocu|       g_hora|g_hora_5|mes_ocu|dia_sem_ocu|   depto_ocu|           mupio_ocu|zona_ocu|sexo_per|edad_per|g_edad_80ymas|g_edad_60ymas|edad_quinquenales|mayor_menor|   tipo_veh|marca_veh|color_veh|modelo_veh|g_modelo_veh|tipo_eve| fall_les|int_o_noint|
+-------+-------+--------+-------------+--------+-------+-----------+------------+--------------------+--------+--------+--------+-------------+-------------+-----------------+-----------+-----------+---------+---------+----------+------------+--------+---------+-----------+
|    nan|     18|       2|00:00 a 05:59|     nan|  Enero|     Sábado|   Guatemala|           Guatemala|      18|  Hombre|Ignorado|          nan|          nan|         Ignor

In [110]:
num_filas = dfhecho.count()
num_filas

100357

Hay 100357 accidentes.

Inciso 5

In [111]:
from pyspark.sql import functions as F

df_anio_dep = dfhecho.groupBy("ano_ocu", "depto_ocu") \
    .count() \
    .orderBy("ano_ocu", "depto_ocu")

display(df_anio_dep) 


DataFrame[ano_ocu: string, depto_ocu: string, count: bigint]

Inciso 6

In [112]:
# Filtramos solo los accidentes del año 2023
df_2023 = dfhecho.filter(F.col("ano_ocu") == "2023")

# Agrupamos por el día de la semana
df_accidentes_dia = df_2023.groupBy("dia_sem_ocu") \
    .count() \
    .orderBy(F.desc("count"))

display(df_accidentes_dia)  # en Databricks selecciona gráfico de columnas


DataFrame[dia_sem_ocu: string, count: bigint]

Inciso 7

In [113]:
from pyspark.sql import functions as F

# Filtrar solo los accidentes del municipio de Guatemala
df_gt = dfhecho.filter(F.col("mupio_ocu") == "Guatemala")

# Agrupar por hora_ocu (ya es de 0 a 23)
df_horas = df_gt.groupBy("hora_ocu") \
    .count() \
    .orderBy("hora_ocu")

display(df_horas)  # en Databricks seleccionas "histograma"


DataFrame[hora_ocu: string, count: bigint]

Inciso 8

Inciso 9

Inciso 10

Inciso 11

Inciso 12

Inciso 13

Inciso 14

Inciso 15

Inciso 16

Inciso 17